lib/psych/visitors/yaml_tree.rb

```diff
@@ -310,19 +310,20 @@ def visit_String o
           style = Nodes::Scalar::LITERAL
           plain = false
           quote = false
-        elsif o =~ /\n/
-          style = Nodes::Scalar::LITERAL
         elsif o == '<<'
           style = Nodes::Scalar::SINGLE_QUOTED
           tag   = 'tag:yaml.org,2002:str'
           plain = false
           quote = false
+        elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
+          style = Nodes::Scalar::LITERAL
+        elsif o.length > line_width
```

Comment:
> It will be IMHO even better for performance than calling lambda for every string.

I'm not sure about that.  If you make the change you've proposed above, you'll be doing 2 comparisons, `@line_width ||=`, and `o.length > line_width`, plus an ivar lookup and 2 method calls, `line_width` and `String#length`.  All these will happen in the common case where someone doesn't provide a line_width, as well as when someone provides a line_width.

What I'm proposing would be an ivar lookup and an essentially noop lambda call in the best case, and the same calculations that you have in the worst case.